rfd/0038-parquet-and-events.md

```diff
@@ -0,0 +1,139 @@
+---
+authors: Joel Wejdenstal (jwejdenstal@goteleport.com)
+state: draft
+---
+
+# RFD 38 - Parquet and a new event architecture
+
+## What
+
+An overhaul/extension of the current way we handle events to accomodate new requirements
+- Natively export to Parquet on-the-fly [#6160](https://github.com/gravitational/teleport/issues/6160)
+- Support encryption of Parquet files and session recordings when stored on cloud services [#6311](https://github.com/gravitational/teleport/issues/6311)
+- Perform advanced queries in real-time on events in an efficient manner [#7724](https://github.com/gravitational/teleport/issues/7724)
+
+## Why
+
+Our requirements have changed and customers want these capabilities to improve integration into their systems and to comply with security regulations.
+
+[#6160](https://github.com/gravitational/teleport/issues/6160) will make us compatible with SIEM and analytics tools which many of our customers use. This is crucial to make events useful on a larger scale.
+
+[#6311](https://github.com/gravitational/teleport/issues/6311) is required because many do not trust the built-in at-rest encryption of S3 want additional security in the form of client side or AWS KMS encryption.
+
+[#7714](https://github.com/gravitational/teleport/issues/7724) is cruicial to enable new views into data from the Teleport webapp and improve the efficiency of existing types of queries.
+
+## Details
+
+This section will go through the requirements, the different architectural and technical solutions I've evaluated, and finally what a sustainable path forward is that fulfills these requirements.
+
+Each section attempts to address one of the concerns outlined above.
+
+### Parquet export [#6160](https://github.com/gravitational/teleport/issues/6160)
+
+This requirement calls for a way to continously export events into Parquet formatted columnar files and periodically uploaded for storage on cloud storage platforms such as AWS S3 and GCP Cloud Storage.
+
+The current events system is structured as the event driver interface `IAuditLog`, implemented for each backend with standardized operations for upload, and query. The simplest way to implement the above feature would be to utilize the `MultiLog` splitter which allows fan-out to multiple backends for upload operations. A new backend implementing `IAuditLog` is to be implemented for Parquet that only supports upload/similar operations and is to be used in tandem with a backend that supports queries such as `DynamoDB`.
+
+When enabled, events will fan-out to this backend and be saved into Parquet files and periodically uploaded after a short timeframe or an event threshold is reached. When AWS S3 is used as the backend cloud storage, AWS Kinesis Firehose should be used to improve the perfromance and reliability of this system by streaming the events from the driver onto a Firehose stream which will take care of the batching and writing to Parquet files.
```

Comment:
I like the idea of batching and optimizing using queue. We need to consider on-prem use-cases when there is no kinesis. Remember we have discussed the format when the system can write multipart S3 parts that are parquet-compatible? Is that possible with parquet format?

Reply:
From my understanding Parquet is columnar which makes this hard. That said if we want Parquet export you can solve this by just creating many smaller Parquet files instead. That said please read below due to the fact that we can't query Parquet so an actual reason for implementing this Parquet driver in the first place is up in the air.